src/net/connection/connection_pool.rs

```diff
@@ -0,0 +1,95 @@
+use super::{ConnectionTimeout, VirtualConnection};
+use error::{NetworkError, Error,Result};
+
+use std::thread;
+use std::time::Duration;
+use std::sync::{Arc, Mutex, RwLock, RwLockWriteGuard};
+use std::net::{SocketAddr};
+use std::collections::HashMap;
+
+// Default timeout of a specific client
+const TIMEOUT_DEFAULT: ConnectionTimeout = 10;
+
+// Default time between checks of all clients for timeouts in seconds
+const TIMEOUT_POLL_INTERVAL: u64 = 1;
+
+/// This is an pool of virtual connections (connected) over UDP.
+pub struct ConnectionPool
+{
+    timeout: ConnectionTimeout,
+    connections: Arc<RwLock<HashMap<SocketAddr, Arc<RwLock<VirtualConnection>>>>>,
+    sleepy_time: Duration,
+    poll_interval: Duration,
+}
+
+impl ConnectionPool {
+    pub fn new() -> ConnectionPool
+    {
+        let sleepy_time = Duration::from_secs(TIMEOUT_DEFAULT);
+        let poll_interval = Duration::from_secs(TIMEOUT_POLL_INTERVAL);
+
+        ConnectionPool { timeout: TIMEOUT_DEFAULT, connections: Arc::new(RwLock::new(HashMap::new())), sleepy_time, poll_interval }
+    }
+
+    /// Set the timeout before an client will be seen as disconnected.
+    pub fn set_timeout(&mut self, timeout: ConnectionTimeout)
+    {
+        self.timeout = timeout;
+    }
+
+    /// Insert connection if it does not exists.
+    pub fn insert_if_not_exists(&mut self, addr: &SocketAddr) -> Result<Arc<RwLock<VirtualConnection>>> {
+        let mut lock = self
+            .connections
+            .write()
+            .map_err(|_| NetworkError::AddConnectionToManagerFailed)?;
+
+        let connection = lock
+            .entry(*addr)
+            .or_insert_with(|| Arc::new(RwLock::new(VirtualConnection::new(*addr))));
+
+        Ok(connection.clone())
+    }
+
+    /// Start loop that will detect if connections will are disconnected.
+    ///
+    /// This function starts a background thread that does the following:
+    /// 1. Gets a read lock on the HashMap containing all the connections
+    /// 2. Iterate through each one
+    /// 3. Check if the last time we have heard from them (received a packet from them) is greater than the amount of time considered to be a timeout
+    /// 4. If they have timed out, send a notification up the stack
+    pub fn start_time_out_loop(&self) -> Result<thread::JoinHandle<()>>
+    {
+        let connections = self.connections.clone();
+        let poll_interval = self.poll_interval;
+
+        Ok(thread::Builder::new()
+            .name("check_for_timeouts".into())
+            .spawn(move ||
+                loop {
+                    match connections.write() {
```

Comment:
I don't think you need a `write` lock here? Since, you are just reading from the hashmap?